### PR TITLE
Error out of default_collate for lists of unequal size

### PR DESCRIPTION
**Pull request by choidongyeon**
Fix issue https://github.com/pytorch/pytorch/issues/23141#

In the below example ```default_collate``` collates each element of the list. Since the second element isn't present in all samples, it is discarded:
```
from torch.utils.data import Dataset
from torch.utils.data import DataLoader
import numpy as np

class CustomDataset(Dataset):
    def __len__(self):
        return 2

    def __getitem__(self, idx):
        tmp = {
            "foo": np.array([1, 2, 3]),
            "bar": ["X"] * (idx+1),
        }

        return tmp

training = CustomDataset()

for batch in DataLoader(training, batch_size=2):
    print(batch)
```
Yields 
```
{
  'foo': tensor(
    [
      [1, 2, 3],
      [1, 2, 3]
    ]
  ),
  'bar': [
      ('X', 'X'),
    ]
}
```

Based on discussion in the issue, it seems the best course of action is to error out in this case. This seems consistent with what is done for tensor elements, as seen in [TensorShape.cpp line 1066](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorShape.cpp#L1060) which is called when ```torch.stack``` is called. In this PR, I introduce a similar message to error out for lists.

@SsnL

### CODE REVIEW

**Review by ssnl — COMMENTED** [review]
could you add a test in `test_dataloader.py` too? something that asserts default_collate raises is fine!

File: torch/utils/data/_utils/collate.py [hunk 1]
```diff
@@ -75,6 +75,10 @@ def default_collate(batch):
     elif isinstance(elem, tuple) and hasattr(elem, '_fields'):  # namedtuple
         return elem_type(*(default_collate(samples) for samples in zip(*batch)))
     elif isinstance(elem, container_abcs.Sequence):
+        it = iter(batch)
+        elem_size = len(next(it))
+        assert all(len(elem) == elem_size for elem in it), \
```
Comment (SsnL) [inline torch/utils/data/_utils/collate.py:80]: Thanks for updating with the test! One final small thing. Generally we don't really throw `AssertionError` (unless it is violation of an invariance which signals a PyTorch bug). Could you change this to throw `RuntimeError` please? Thanks!

Comment (choidongyeon) [reply torch/utils/data/_utils/collate.py]: Wow, that was a quick review. Addressed in newest update!

**Review by ssnl — APPROVED** [review]
Thank you :)

**Comment by choidongyeon** [comment]
Thanks for taking a look @SsnL!

**Pull request closed without merging**